pkg/dockerregistry/server/auth.go

```diff
@@ -251,56 +256,65 @@ func (ac *AccessController) Authorized(ctx context.Context, accessRecords ...reg
 					continue
 				}
 				if err := verifyPruneAccess(ctx, osClient); err != nil {
-					return nil, ac.wrapErr(err)
+					return nil, ac.wrapErr(ctx, err)
 				}
 				verifiedPrune = true
 			default:
-				return nil, ac.wrapErr(ErrUnsupportedAction)
+				return nil, ac.wrapErr(ctx, ErrUnsupportedAction)
 			}
 		default:
-			return nil, ac.wrapErr(ErrUnsupportedResource)
+			return nil, ac.wrapErr(ctx, ErrUnsupportedResource)
 		}
 	}
 
 	// deal with any possible cross-mount errors
 	for namespaceAndName, err := range possibleCrossMountErrors {
-		// If we have no push requests, this can't be a cross-mount request, so error
-		if len(pushChecks) == 0 {
-			return nil, err
-		}
-		// If we also requested a push to this ns/name, this isn't a cross-mount request, so error
-		if pushChecks[namespaceAndName] {
-			return nil, err
+		// If we have no push requests, this can't be a cross-mount request, so error.
+		// Error also in case of a push being requested this ns/name, which isn't a cross-mount request.
+		if len(pushChecks) == 0 || pushChecks[namespaceAndName] {
+			context.GetLogger(ctx).Error(err.Error())
+			return nil, ac.wrapErr(ctx, err)
 		}
 	}
 
 	// Conditionally add auth errors we want to handle later to the context
 	if !possibleCrossMountErrors.Empty() {
-		context.GetLogger(ctx).Debugf("Origin auth: deferring errors: %#v", possibleCrossMountErrors)
-		ctx = WithDeferredErrors(ctx, possibleCrossMountErrors)
+		errs := deferredErrors{}
+		for repository, err := range possibleCrossMountErrors {
+			context.GetLogger(ctx).Infof("Origin auth: deferring authorization error for repository %q: %v", repository, err)
+			errs[repository] = ac.wrapErr(ctx, err)
+		}
+		ctx = WithDeferredErrors(ctx, errs)
 	}
 	// Always add a marker to the context so we know auth was run
 	ctx = WithAuthPerformed(ctx)
 
 	return WithUserClient(ctx, osClient), nil
 }
 
-func getNamespaceName(resourceName string) (string, string, error) {
-	repoParts := strings.SplitN(resourceName, "/", 2)
-	if len(repoParts) != 2 {
-		return "", "", ErrNamespaceRequired
-	}
-	ns := repoParts[0]
-	if len(ns) == 0 {
-		return "", "", ErrNamespaceRequired
-	}
-	name := repoParts[1]
-	if len(name) == 0 {
-		return "", "", ErrNamespaceRequired
+// wrapErr wraps errors related to authorization in an authChallenge error that will present a WWW-Authenticate challenge response
+func (ac *AccessController) wrapErr(ctx context.Context, err error) error {
+	switch err.(type) {
+	case *errAccessDenied:
+		context.GetLogger(ctx).Error(err.Error())
```

Comment:
I would not expect wrapErr to be logging... doing so means that we could still get nasty error messages logged for benign cases like this: https://github.com/openshift/origin/pull/9819/commits/c4f5056a29c8fda4794572327572e636d927d0a7#diff-4fcaa728857c0df2a0e21105899aedb3R285